File: 조은정/Lv.2/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <stack>
using namespace std;

int solution(string s)
{
    int answer = s.size(), chk;
    char c;
    stack<char> st;

    for (int i = 0; i < s.size(); ++i)
    {
        chk = 1;
        for (int j = i; j < s.size(); ++j)
        {
            c = s[j];
            if (c == ')' || c == '}' || c == ']')
            {
                if (!st.size() || (c == ')' && st.top() != '(') ||
                    (c == '}' && st.top() != '{') || (c == ']' && st.top() != '['))
                {
                    chk = 0;
                    --answer;
                    break;
                }
                st.pop();
            }
            else
                st.push(c);
        }
        for (int j = 0; j < i && chk; ++j)
        {
            c = s[j];
            if (c == ')' || c == '}' || c == ']')
            {
                if (!st.size() || (c == ')' && st.top() != '(') ||
                    (c == '}' && st.top() != '{') || (c == ']' && st.top() != '['))
                {
                    --answer;
                    break;
                }
                st.pop();
            }
            else
                st.push(c);
        }
        if (chk && st.size())
            --answer;
        st = {};
    }

    return answer;
}

File: 신창호/Lv2/괄호 회전하기.java
import java.util.*;
class Solution {
    public int solution(String s) {
        // 완전탐색 10^3
        // []
        int answer = 0;
        Stack<Character> stack;
        boolean check;
        int index;
        for(int start = 0;start < s.length(); start++){
            stack = new Stack<>();
            index = 0;
            check = true;
            while(index < s.length()){
                int l = start+index;
                if(l >= s.length()){
                    l -= s.length();
                }
                Character c= s.charAt(l);
                index++;
                if(c == '[' ||c == '{' || c == '(' ){
                    stack.add(c);
                }
                if(c == ']'){
                    if(!stack.isEmpty() && stack.peek() == '['){
                        stack.pop();
                        continue;
                    }
                    check = false;
                    index = s.length(); // 종료
                }
                if(c == '}'){
                    if(!stack.isEmpty() && stack.peek() == '{'){
                        stack.pop();
                        continue;
                    }
                    check = false;
                    index = s.length(); // 종료
                }
                if(c == ')'){
                    if(!stack.isEmpty() && stack.peek() == '('){
                        stack.pop();
                        continue;
                    }
                    check = false;
                    index = s.length(); // 종료
                }


            }
            if(stack.isEmpty() && check){
                answer++;
            }
        }
        return answer;
    }
}